### DISCUSSION TITLE
app can't find any command

### DISCUSSION THREAD

**Discussion by lipeyey** [Rapptz/discord.py#9123] [Q&A]
im creating a bot, but it cant find any command that is in a cog. i've tried everything, here's my code

main:
import discord
import os
import asyncio

from discord.ext import commands

TokenFile = open("./data/Token.txt", "r") 
TOKEN = TokenFile.read()

OWNERID = 755431789987758140

intents = discord.Intents.all()

bot = commands.Bot(command_prefix="!", help_command=None,intents=intents)

async def load():
    for filename in os.listdir("./cogs"):
               if filename.endswith(".py") and not filename.startswith('__'):
                      await bot.load_extension("cogs.{0}".format(filename[:-3]))

@bot.event
async def on_ready():
    print('Logged in as')
    print(bot.user.name)
    print(bot.user.id)
    print('------')

@bot.command(name='ping')
async def ping(ctx):
        await ctx.send(f'**Pong!** Latency: {round(bot.latency * 1000)}ms')

bot.run(str(TOKEN))

cog:
import random

import discord
from discord.ext import commands
from discord import slash_command

intents = discord.Intents.all()
bot = commands.Bot(command_prefix="!", intents=intents)

sites = ["https://longdogechallenge.com/",
         "http://heeeeeeeey.com/",
         "http://corndog.io/",
         "https://mondrianandme.com/",
         "https://puginarug.com",
         "https://alwaysjudgeabookbyitscover.com",
         "https://thatsthefinger.com/",
         "https://cant-not-tweet-this.com/",
         "http://eelslap.com/",
         "http://www.staggeringbeauty.com/",
         "http://burymewithmymoney.com/",
         "https://smashthewalls.com/",
         "https://jacksonpollock.org/",
         "http://endless.horse/",
         "https://www.trypap.com/",
         "http://www.republiquedesmangues.fr/",
         "http://www.movenowthinklater.com/",
         "http://www.partridgegetslucky.com/",
         "http://www.rrrgggbbb.com/",
         "http://beesbeesbees.com/",
         "http://www.koalastothemax.com/",
         "http://www.everydayim.com/",
         "http://randomcolour.com/",
         "http://cat-bounce.com/",
         "http://chrismckenzie.com/",
         "https://thezen.zone/",
         "http://hasthelargehadroncolliderdestroyedtheworldyet.com/",
         "http://ninjaflex.com/",
         "http://ihasabucket.com/",
         "http://corndogoncorndog.com/",
         "http://www.hackertyper.com/",
         "https://pointerpointer.com",
         "http://imaninja.com/",
         "http://drawing.garden/",
         "http://www.ismycomputeron.com/",
         "http://www.nullingthevoid.com/",
         "http://www.muchbetterthanthis.com/",
         "http://www.yesnoif.com/",
         "http://lacquerlacquer.com",
         "http://potatoortomato.com/",
         "http://iamawesome.com/",
         "https://strobe.cool/",
         "http://www.pleaselike.com/",
         "http://crouton.net/",
         "http://corgiorgy.com/",
         "http://www.wutdafuk.com/",
         "http://unicodesnowmanforyou.com/",
         "http://chillestmonkey.com/",
         "http://scroll-o-meter.club/",
         "http://www.crossdivisions.com/",
         "http://tencents.info/",
         "https://boringboringboring.com/",
         "http://www.patience-is-a-virtue.org/",
         "http://pixelsfighting.com/",
         "http://isitwhite.com/",
         "https://existentialcrisis.com/",
         "http://onemillionlols.com/",
         "http://www.omfgdogs.com/",
         "http://oct82.com/",
         "http://chihuahuaspin.com/",
         "http://www.blankwindows.com/",
         "http://dogs.are.the.most.moe/",
         "http://tunnelsnakes.com/",
         "http://www.trashloop.com/",
         "http://www.ascii-middle-finger.com/",
         "http://spaceis.cool/",
         "http://www.donothingfor2minutes.com/",
         "http://buildshruggie.com/",
         "http://buzzybuzz.biz/",
         "http://yeahlemons.com/",
         "http://wowenwilsonquiz.com",
         "https://thepigeon.org/",
         "http://notdayoftheweek.com/",
         "http://www.amialright.com/",
         "http://nooooooooooooooo.com/",
         "https://greatbignothing.com/",
         "https://zoomquilt.org/",
         "https://dadlaughbutton.com/",
         "https://www.bouncingdvdlogo.com/",
         "https://remoji.com/",
         "http://papertoilet.com/", ]

ois_do_bot = ['oi', 'eae', 'eai', 'opa', 'epa', 'yo', 'alou', 'ola', 'Olá', 'epa', 'manda a braba', 'fala fi',
              'opa, bom?', 'oi, como estais?', 'chamou?']

respostas =  ['Não respondo a isso', 'Sim', 'As vezes', 'Não', 'Claro', 'NUNCA!', 'Um dia talvez',
                   'A resposta está dentro de vc', 'Mais ou menos', 'Podia ser pior']

class Fun(commands.Cog):
    """
    fun commands to make you happy :grin:
    """
    def __init__(self, bot):
        self.bot = bot
        self._last_member = None

    @slash_command(name='pergunta', aliases=['d'])
    async def pergunta(self, ctx):
        """just ask something and it will answer you"""
        await ctx.send(random.choice(respostas))

    @commands.command(name='oi', aliases=[ 'eae', 'eai', 'opa', 'yo', 'ei', 'alou', 'ola', 'Olá', 'epa'])
    async def hi(self, ctx):
        """
        just send hi if you are feeling alone
        """
        await ctx.send(random.choice(ois_do_bot))

    @commands.command(name='dice', aliases=['roll'])
    async def dice(self, ctx):
        """
        roll a 6 sided dice
        """
        await ctx.send(random.randint(1, 6))

    @commands.command(name="site", aliases=["link"])
    async def site(self, ctx):
        """
        send you a funny site link
        """

        await ctx.send(random.choice(sites))

bot.add_cog(Fun(bot))

**Comment by sparrow001** [comment]
It is slightly hard to read because of the weird formatting, but it seems you don't have a cog setup function. You can't just do bot.add_cog, you have to wrap it in a setup function. You can find an example [here](https://gist.github.com/pinheaded/ac880263084a65b8d2f769c1fa46a3d7)

**Comment by CoderJoshDK** [comment]
There are a lot of things going on here. And it is really hard to read because you didn't use code blocks or have things indented.  
But, from what I could tell, for one, you are not loading in the exertions. You have a function for that:
```py
async def load():
  for filename in os.listdir("./cogs"):
    if filename.endswith(".py") and not filename.startswith('__'):
      await bot.load_extension("cogs.{0}".format(filename[:-3]))
```
But this function is never called anywhere. Make sure to actually call the function.

Better still, you should look into making a subclass of the bot so you can [override setup_hook](https://discordpy.readthedocs.io/en/latest/migrating.html#migrating-2-0-commands-extension-cog-async).

On another note, it looks like you are making another bot in your cog"?" When you load the extension, it will handle passing the bot along if you have a section like:
```py
async def setup(bot: commands.Bot) -> None:
  await bot.add_cog(MyCog(bot))
```
A great read-up you should look into is [Slash Commands and you by AbstractUmbra ](https://gist.github.com/AbstractUmbra/a9c188797ae194e592efe05fa129c57f). It should have a lot of good examples.

**Comment by externref** [comment]
thats not how application commands in discord.py work.
if you're not getting any errors you have some other library ( probably a fork ?) using the discord namespace installed, or one that monkeypatches the library.
consider removing any such libraries and rewriting your code according to dpy implementation
Docs: https://discordpy.readthedocs.io/en/stable/interactions/api.html#application-commands
Example:
https://github.com/Rapptz/discord.py/blob/master/examples/app_commands/basic.py

you can find better guidance in the official discord (.gg/dpy)

**Comment by felipeyx** [comment]
There are two problems in the shown structure.

First, do not create another commands.Bot inside the cog file. There should normally be only one bot instance.

A cog should look roughly like:

class MyCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command()
    async def randomsite(self, ctx):
        await ctx.send("example")


async def setup(bot):
    await bot.add_cog(MyCog(bot))

Second, you defined a function that loads extensions, but make sure it is actually called before the bot starts.

A common pattern for discord.py 2.x is:

async def main():
    async with bot:
        await bot.load_extension("cogs.example")
        await bot.start(TOKEN)

asyncio.run(main())

You can verify loading with:

print(bot.extensions)
print([c.name for c in bot.commands])

Creating another Bot inside the cog means its commands are registered to that unused bot rather than the one connected to Discord.
